BioCro/DARPA/set_MA_trt.R

```diff
@@ -0,0 +1,108 @@
+
+  #create new trait.mcmc of combined random effects
+  new.trait.mcmc <- mapply(FUN = RE_combine, mc = trait.mcmc, trt = trt.match, 
+                           SIMPLIFY = FALSE)
+  
+  #identify target site and treatment
+  target.site <- settings$run$site$id
+  target.trt <- settings$meta.analysis$treatment
+
+  #label target site/treatment "beta.o"
+  final.trait.mcmc <- mapply(FUN = rename_cols, trait.mcmc = new.trait.mcmc, trt = trt.match,
+                             MoreArgs = list(target.site = target.site, target.trt = target.trt),
+                             SIMPLIFY = FALSE)
+  
+  #rename as trait.mcmc and save as "trait.mcmc.Rdata", 2 places
+  trait.mcmc <- final.trait.mcmc
+  save(trait.mcmc, file = file.path(settings$database$dbfiles, "posterior", postid, "trait.mcmc.Rdata"))
+  save(trait.mcmc, file = file.path(settings$pfts$pft$outdir, "trait.mcmc.Rdata"))
+}
+
+#Returns unique combination of site, trt, and ghs for matching purposes
+collapse <- function(jagged){
+  match <- unique(jagged[,c("site", "trt", "ghs", 
+                          "site_id", "treatment_id", "greenhouse", 
+                          "trt_name", "trt_num")])
+  colnames(match)[colnames(match) %in% c("trt", "trt_num")] <- c("trt_control", "trt")
+  row.names(match) <- NULL
+  return(match)
+}
+
+
+#Combines correct columns of RE for each treatment
+RE_combine <- function(mc, trt){
+  #create table of relevant RE for each treatment (row)
+  trt <- trt[, c("ghs", "site", "trt")]
+  #empty matrix of column indices for matching column names
+  col_ind <- matrix(NA, nrow = nrow(trt), ncol = ncol(trt))
+  #existing column names
+  cnames <- colnames(mc[[1]])
+  #fill col_ind with the column index of the matching column names
+  for(i in 1:nrow(trt)){
+    for(j in 1:ncol(trt)){
+      index <- which(cnames == paste0("beta.", colnames(trt)[j], "[", trt[i,j], "]"))
+      col_ind[i,j] <- if(length(index) == 1){
+        index
+      } else {
+          NA
+        }
+    }
+  }
+  #add index for "beta.o", which is universal across treatments
+  col_ind <- cbind(col_ind, rep(which(cnames == "beta.o"), nrow(col_ind)))
+  
+  #create empty mcmc.list object
+  chain <- list(mcmc(matrix(NA, nrow = nrow(mc[[1]]), ncol = nrow(trt))))
+  out <- rep(chain, length(mc))
+  #for each chain and treatment, combine the relevant posterior RE from trait.mcmc
+  for(c in 1:length(mc)){
+    for(t in 1:nrow(trt)){ # number of treatments
+      out[[c]][,t] <- rowSums(mc[[c]][ ,col_ind[t,]], na.rm = T)
+    }
+  }
+  return(mcmc.list(out))
+}
+
+#Renames target treatment column with "beta.o"
+rename_cols <- function(trait.mcmc, trt, target.site, target.trt){
+  #index of target treatment
+  ind <- which(trt$site_id == target.site & trt$trt_name == target.trt)
+  
+  #for each chain, add trt_name as column names; then sub out the traget treatment column name with "beta.o"
+  for(c in 1:length(trait.mcmc)){
+    colnames(trait.mcmc[[c]]) <- trt$trt_name
```

Comment:
Are these for sure always going to be in the same order?

Reply:
See line 99, where ind is evaluated as the correct column index.